add force_set_lock, force_remove_lock call to currency-adapter

At the beginning, we just check if the `AssetId` is 0 OR 1, if not then return Error. If yes, we can then use pallet_balances's set_lock & remove_lock

@GopherJ I am getting below error if I move into dispatchable


```
    Checking pallet-currency-adapter v1.7.7 (/Users/ayushkumarmishra/workspace/parallel/pallets/currency-adapter)
error[E0277]: the trait bound `frame_support::traits::WithdrawReasons: scale_info::TypeInfo` is not satisfied
  --> pallets/currency-adapter/src/lib.rs:46:1
   |
46 |   #[frame_support::pallet]
   |   ^-----------------------
   |   |
   |  _in this procedural macro expansion
   | |
47 | | pub mod pallet {
48 | |     use super::*;
49 | |     use frame_support::traits::LockableCurrency;
...  |
76 | |
77 | |     #[pallet::call]
   | |__________________^ the trait `scale_info::TypeInfo` is not implemented for `frame_support::traits::WithdrawReasons`
   |
   = note: this error originates in the derive macro `frame_support::scale_info::TypeInfo` (in Nightly builds, run with -Z macro-backtrace for more info)

```


Can we use hardcoded `WithdrawReasons` like https://github.com/paritytech/substrate/blob/master/frame/democracy/src/lib.rs#L1128 
https://github.com/paritytech/substrate/blob/master/frame/elections-phragmen/src/lib.rs#L347

--- a/pallets/currency-adapter/src/lib.rs
+++ b/pallets/currency-adapter/src/lib.rs
@@ -22,8 +22,11 @@
 
 pub use pallet::*;
 
+use frame_support::traits::LockableCurrency;
+use frame_support::traits::{LockIdentifier, WithdrawReasons};
 use frame_support::{
     dispatch::DispatchResult,
+    pallet_prelude::*,
     traits::{
         tokens::{
             fungible::{Inspect, Mutate, Transfer},
@@ -44,6 +47,7 @@ type BalanceOf<T> =
 #[frame_support::pallet]
 pub mod pallet {
     use super::*;
+    use frame_support::traits::LockableCurrency;
 
     #[pallet::config]
     pub trait Config: frame_system::Config {
@@ -53,7 +57,8 @@ pub mod pallet {
 
         type Balances: Inspect<Self::AccountId, Balance = Balance>
             + Mutate<Self::AccountId, Balance = Balance>
-            + Transfer<Self::AccountId, Balance = Balance>;
+            + Transfer<Self::AccountId, Balance = Balance>
+            + LockableCurrency<Self::AccountId, Balance = Balance, Moment = Self::BlockNumber>;
 
         #[pallet::constant]
         type GetNativeCurrencyId: Get<AssetIdOf<Self>>;
@@ -62,8 +67,44 @@ pub mod pallet {
     #[pallet::pallet]
     pub struct Pallet<T>(_);
 
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Not a native token
+        NotANativeToken,
+    }
+
     #[pallet::call]
     impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> Pallet<T> {
+    pub fn force_set_lock(
+        asset: AssetIdOf<T>,
+        id: LockIdentifier,
+        who: T::AccountId,
+        amount: BalanceOf<T>,
+        reasons: WithdrawReasons,
+    ) -> DispatchResult {
+        ensure!(
+            asset == T::GetNativeCurrencyId::get(),
+            Error::<T>::NotANativeToken
+        );
+        T::Balances::set_lock(id, &who, amount, reasons);
+        Ok(())
+    }
+
+    pub fn force_remove_lock(
+        asset: AssetIdOf<T>,
+        id: LockIdentifier,
+        who: T::AccountId,
+    ) -> DispatchResult {
+        ensure!(
+            asset == T::GetNativeCurrencyId::get(),
+            Error::<T>::NotANativeToken
+        );
+        T::Balances::remove_lock(id, &who);
+        Ok(())
+    }
 }
 
 impl<T: Config> Inspects<T::AccountId> for Pallet<T> {

--- a/pallets/currency-adapter/src/lib.rs
+++ b/pallets/currency-adapter/src/lib.rs
@@ -22,7 +22,6 @@
 
 pub use pallet::*;
 
-use frame_support::traits::LockableCurrency;
 use frame_support::traits::{LockIdentifier, WithdrawReasons};
 use frame_support::{
     dispatch::DispatchResult,
@@ -48,6 +47,8 @@ type BalanceOf<T> =
 pub mod pallet {
     use super::*;
     use frame_support::traits::LockableCurrency;
+    use frame_system::ensure_root;
+    use frame_system::pallet_prelude::OriginFor;
 
     #[pallet::config]
     pub trait Config: frame_system::Config {
@@ -74,36 +75,40 @@ pub mod pallet {
     }
 
     #[pallet::call]
-    impl<T: Config> Pallet<T> {}
-}
+    impl<T: Config> Pallet<T> {
+        #[pallet::weight(10_000)]
+        pub fn force_set_lock(
+            origin: OriginFor<T>,
+            asset: AssetIdOf<T>,
+            id: LockIdentifier,
+            who: T::AccountId,
+            amount: BalanceOf<T>,
+            reasons: WithdrawReasons,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            ensure!(
+                asset == T::GetNativeCurrencyId::get(),
+                Error::<T>::NotANativeToken
+            );
+            T::Balances::set_lock(id, &who, amount, reasons);
+            Ok(())
+        }
 
-impl<T: Config> Pallet<T> {
-    pub fn force_set_lock(
-        asset: AssetIdOf<T>,
-        id: LockIdentifier,
-        who: T::AccountId,
-        amount: BalanceOf<T>,
-        reasons: WithdrawReasons,
-    ) -> DispatchResult {
-        ensure!(
-            asset == T::GetNativeCurrencyId::get(),
-            Error::<T>::NotANativeToken
-        );
-        T::Balances::set_lock(id, &who, amount, reasons);
-        Ok(())
-    }
-
-    pub fn force_remove_lock(
-        asset: AssetIdOf<T>,
-        id: LockIdentifier,
-        who: T::AccountId,
-    ) -> DispatchResult {
-        ensure!(
-            asset == T::GetNativeCurrencyId::get(),
-            Error::<T>::NotANativeToken
-        );
-        T::Balances::remove_lock(id, &who);
-        Ok(())
+        #[pallet::weight(10_000)]
+        pub fn force_remove_lock(
+            origin: OriginFor<T>,
+            asset: AssetIdOf<T>,
+            id: LockIdentifier,
+            who: T::AccountId,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            ensure!(
+                asset == T::GetNativeCurrencyId::get(),
+                Error::<T>::NotANativeToken
+            );
+            T::Balances::remove_lock(id, &who);
+            Ok(())
+        }
     }
 }
 

--- a/pallets/currency-adapter/src/lib.rs
+++ b/pallets/currency-adapter/src/lib.rs
@@ -22,7 +22,6 @@
 
 pub use pallet::*;
 
-use frame_support::traits::{LockIdentifier, WithdrawReasons};
 use frame_support::{
     dispatch::DispatchResult,
     pallet_prelude::*,
@@ -32,7 +31,7 @@ use frame_support::{
             fungibles::{Inspect as Inspects, Mutate as Mutates, Transfer as Transfers},
             DepositConsequence, WithdrawConsequence,
         },
-        Get,
+        Get, LockIdentifier, WithdrawReasons,
     },
 };
 use primitives::{Balance, CurrencyId};
@@ -47,8 +46,7 @@ type BalanceOf<T> =
 pub mod pallet {
     use super::*;
     use frame_support::traits::LockableCurrency;
-    use frame_system::ensure_root;
-    use frame_system::pallet_prelude::OriginFor;
+    use frame_system::{ensure_root, pallet_prelude::OriginFor};
 
     #[pallet::config]
     pub trait Config: frame_system::Config {
@@ -82,15 +80,14 @@ pub mod pallet {
             asset: AssetIdOf<T>,
             id: LockIdentifier,
             who: T::AccountId,
-            amount: BalanceOf<T>,
-            reasons: WithdrawReasons,
+            #[pallet::compact] amount: BalanceOf<T>,
         ) -> DispatchResult {
             ensure_root(origin)?;
             ensure!(
                 asset == T::GetNativeCurrencyId::get(),
                 Error::<T>::NotANativeToken
             );
-            T::Balances::set_lock(id, &who, amount, reasons);
+            T::Balances::set_lock(id, &who, amount, WithdrawReasons::all());
             Ok(())
         }
 

